### ISSUE TITLE
Test failure test_set_focus_date_weekstart_6

### ISSUE THREAD

**Issue by lfam** [pimutils/khal#458]
This test fails for me when building khal 0.8.2, with https://github.com/pimutils/khal/commit/f3e8d82118f629950f36035a4c8cb5775746568b applied to the release tarball.

python-icalendar is 3.10.
python-dateutil is 2.5.2.

```
=================================== FAILURES ===================================
_______________________ test_set_focus_date_weekstart_6 ________________________

    def test_set_focus_date_weekstart_6():
    
        with freeze_time('2016-04-10'):
            today = date.today()
            for diff in range(-21, 21, 1):
                frame = CalendarWidget(on_date_change=lambda _: None,
                                       keybindings=keybindings,
                                       on_press=on_press,
                                       firstweekday=6,
>                                      weeknumbers='right')

tests/ui/test_calendarwidget.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
khal/ui/calendarwidget.py:607: in __init__
    self.set_focus_date(self._initial)
khal/ui/calendarwidget.py:624: in set_focus_date
    self.box.set_focus_date(a_day)
khal/ui/calendarwidget.py:297: in set_focus_date
    self.body.set_focus_date(a_day)
khal/ui/calendarwidget.py:371: in set_focus_date
    row, column = self.get_date_pos(a_day)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = CalendarWalker([<DateCColumns from 2016-05-29 to 2016-06-04>, <DateCColumns from 2016-06-05 to 2016-06-11>, <DateCColu...7 to 2016-07-23>, <DateCColumns from 2016-07-24 to 2016-07-30>, <DateCColumns from 2016-07-31 to 2016-08-06>], focus=4)
a_day = FakeDate(2016, 4, 10)

    def get_date_pos(self, a_day):
        """get row and column where `a_day` is located
    
            :type: a_day: datetime.date
            :rtype: tuple(int, int)
            """
        # rough estimate of difference in lines, i.e. full weeks, we might be
        # off by as much as one week though
        week_diff = int((self.focus_date - a_day).days / 7)
        new_focus = self.focus - week_diff
        # in case new_focus is 1 we will later try set the focus to 0 which
        # will lead to an autoprepend which will f*ck up our estimation,
        # therefore better autoprepending anyway, even if it might not be
        # necessary
        if new_focus <= 1:
            self._autoprepend()
            week_diff = int((self.focus_date - a_day).days / 7)
            new_focus = self.focus - week_diff
        for offset in [0, -1, 1]:  # we might be off by a week
            row = new_focus + offset
            try:
                if row >= len(self):
                    self._autoextend()
                column = self[row].get_date_column(a_day)
                return row, column
            except ValueError:
                pass
        # we didn't find the date we were looking for...
>       raise ValueError('something is wrong')
E       ValueError: something is wrong

khal/ui/calendarwidget.py:414: ValueError
=============== 1 failed, 174 passed, 2 xfailed in 39.76 seconds ===============
```

**Comment by geier** [comment]
Strangely, I can confirm this on 0.8.2 but not on master, but the relevant code has not changed....

**Comment by pwnage101** [comment]
Using the latest Guix distribution, I am getting a different result than the one @lfam got.  The upshot: it now fails in a different way, and it has already been reported in #466, and commit 8a80282 fixes it (that commit is not in 0.8.x, but it is in master). It's fascinating to me that the 0.8.x branch even passes tests in Travis :)  

Here's what I did to try to reproduce the test failure in the first comment of this thread:

I updated the Khal package to use version 0.8.2 from pypi, then I added a patch representing f3e8d82.  For completeness I also updated the python-icalendar package to 3.10, just as @lfam did.

The relevant output form the build log ([full output](https://gist.github.com/pwnage101/cb9258375ab40bc9c5071e1a1f1a819e#file-build_log-txt)):
```
starting phase `check'
============================= test session starts ==============================
platform linux -- Python 3.4.3 -- py-1.4.31 -- pytest-2.7.3
rootdir: /tmp/guix-build-khal-0.8.2.drv-0/khal-0.8.2, inifile: 
plugins: cov
collected 176 items

tests/aux_test.py .....................
tests/backend_test.py .........................
tests/cal_display_test.py ....x.
tests/cli_test.py .....x.............
tests/controller_test.py ....
tests/event_test.py .............................
tests/khalendar_aux_test.py ................................
tests/khalendar_test.py ......................
tests/settings_test.py ........
tests/terminal_test.py ...
tests/vtimezone_test.py ...
tests/ui/test_calendarwidget.py ..F
tests/ui/test_widgets.py .

=================================== FAILURES ===================================
_______________________ test_set_focus_date_weekstart_6 ________________________

    def test_set_focus_date_weekstart_6():
    
        with freeze_time('2016-04-10'):
            today = date.today()
            for diff in range(-21, 21, 1):
                frame = CalendarWidget(on_date_change=lambda _: None,
                                       keybindings=keybindings,
                                       on_press=on_press,
                                       firstweekday=6,
>                                      weeknumbers='right')

tests/ui/test_calendarwidget.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
khal/ui/calendarwidget.py:607: in __init__
    self.set_focus_date(self._initial)
khal/ui/calendarwidget.py:624: in set_focus_date
    self.box.set_focus_date(a_day)
khal/ui/calendarwidget.py:297: in set_focus_date
    self.body.set_focus_date(a_day)
khal/ui/calendarwidget.py:371: in set_focus_date
    row, column = self.get_date_pos(a_day)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = CalendarWalker([<DateCColumns from 2016-06-26 to 2016-07-02>, <DateCColumns from 2016-07-03 to 2016-07-09>, <DateCColu...4 to 2016-08-20>, <DateCColumns from 2016-08-21 to 2016-08-27>, <DateCColumns from 2016-08-28 to 2016-09-03>], focus=5)
a_day = FakeDate(2016, 4, 10)

    def get_date_pos(self, a_day):
        """get row and column where `a_day` is located
    
            :type: a_day: datetime.date
            :rtype: tuple(int, int)
            """
        # rough estimate of difference in lines, i.e. full weeks, we might be
        # off by as much as one week though
        week_diff = int((self.focus_date - a_day).days / 7)
        new_focus = self.focus - week_diff
        # in case new_focus is 1 we will later try set the focus to 0 which
        # will lead to an autoprepend which will f*ck up our estimation,
        # therefore better autoprepending anyway, even if it might not be
        # necessary
        if new_focus <= 1:
            self._autoprepend()
            week_diff = int((self.focus_date - a_day).days / 7)
            new_focus = self.focus - week_diff
        for offset in [0, -1, 1]:  # we might be off by a week
            row = new_focus + offset
            try:
                if row >= len(self):
                    self._autoextend()
>               column = self[row].get_date_column(a_day)
E               IndexError: list index out of range

khal/ui/calendarwidget.py:409: IndexError
=============== 1 failed, 173 passed, 2 xfailed in 26.72 seconds ===============
phase `check' failed after 36.1 seconds
```

Here's the dependency closure, excluding things such as GCC: https://gist.github.com/pwnage101/cb9258375ab40bc9c5071e1a1f1a819e#file-closure-txt

Using a combination of `guix environment` and `git bisect`, I was able to determine that cherry picking 8a80282 into the 0.8.x branch fixes the bug which makes this test fail.

**Comment by pwnage101** [comment]
The way I see it, one of two things should happen:

1. we deem this bug to be "not too bad", disable the test, and package 0.8.2 for Guix.
1. backport 8a80282 into 0.8.x and release 0.8.3

Thoughts?

**Comment by geier** [comment]
@pwnage101 thank you for going the extra mile and identifying 8a80282. If you want to get the 0.8 branch into guix as soon as possible, I suggest you disable the test, otherwise wait some days and I'll fix some other bugs in the next few days and we'll release 0.8.3 very soon (preferably before the weekend).

**Comment by pwnage101** [comment]
Makes sense!  I'll hold off on packaging khal 0.8.x, no rush.

**Comment by lfam** [comment]
Hi Troy,

Thanks for paying attention to this!

Feel free to check out the work I did so far on upgrading khal, icalendar, etc:

https://github.com/lfam/guix/commits/contrib-khal

**Comment by geier** [comment]
not actually fixed in master, but in v0.8.x

**Comment by geier** [comment]
released v0.8.3
